### ISSUE TITLE
[Feature request]: Hyper Parameters optimizer

### ISSUE THREAD

**Issue by nicaogr** [tensorflow/tensorflow#12758]
Hi,

Do you plan to add a code in order to optimize automatically the hyper-parameters of deep learning algorithm ? (number of hidden layers, value of the learning rate, type of down-sampling etc).

Best,

**Comment by TomorrowIsAnOtherDay** [comment]
It seems like a paper on NIPS 2016, which used Reinforcement Learning to learn the network structure, but it cost a lot of sources.

**Comment by nicaogr** [comment]
Yes, it is a state of the art method but it could be really useful to be able to build with this kind of computation block in order to obtain better deep learning model. 

It consisted in doing machine learning on the machine learning hyper parameters.

**Comment by yaroslavvb** [comment]
This is a fairly vague feature request since there are so many hyper-parameter optimization methods. The most commonly used one is the learning rate optimization. You launch several runs in parallel, with different learning rates in increments of 1.5, and see which one works better in a reasonable time.

At Google 1.5 years ago there were a few hyper-parameter optimizers implemented for TensorFlow, and I can't recall any of them being too practical. Couple of issues I remember

1. to run hyper-parameter candidates in parallel needs extra infrastructure setup. Imagine your model trains with 50 replicas to train, a hyper-parameter optimizer needs to launch x copies of 50 replicas.

2. hyper-parameter optimizers tend to have their own hyper-parameters that they are sensitive too. For instance, you have to decide how long to train your model with new parameters. Shrinking the learning rate is known to increase accuracy in short term, but will harm performance in the long term. However, training each candidate to converge is not practical for hyper-parameter optimizer.

I think it would make sense to request a specific hyper-parameter optimizer that works well for a specific application -- ie, maybe a specific algorithm that is known to achieve good results on for standard image recognition tasks.

**Comment by jart** [comment]
I agree with @yaroslavvb. Automating this sort of thing is a broad problem space that can be tackled in a variety of ways. We're open to feature requests in this area, but it helps when they're as specific as possible. Anyone with ideas like that, which could help TensorFlow improve, please feel free to open a new issue.
